nntrainer/tensor/cuda_operations/gemm_int4_cpu.cpp

```diff
@@ -0,0 +1,175 @@
+        unsigned int offset_in_group = k % quantization_group_size;
+
+        // Input is stored block-wise: group_id * group_size + offset
+        unsigned int input_idx =
+          global_group_id * quantization_group_size + offset_in_group;
+
+        int8_t q_val = q_input[input_idx];
+
+        // Get input scale
+        // scales[group_id * 2] = scale, scales[group_id * 2 + 1] =
+        // offset/unused
+        float in_scale = compute_fp16_to_fp32(in_scales[global_group_id * 2]);
+
+        float in_val = static_cast<float>(q_val) * in_scale;
+        float w_val = dequantized_weight_row[k];
+
+        sum += in_val * w_val;
+      }
+
+      output[m * N + n] = sum;
+    }
+  }
+}
+
+void gemm_int4_cpu_packed_block(const void *input, const void *weights,
+                                const void *scales, const void *input_scales,
+                                float *output, unsigned int M, unsigned int N,
+                                unsigned int K,
+                                unsigned int quantization_group_size) {
+  const int8_t *q_input = static_cast<const int8_t *>(input);
+  const uint8_t *q_weights = static_cast<const uint8_t *>(weights);
+  const uint16_t *w_scales = static_cast<const uint16_t *>(scales);
+  const uint16_t *in_scales = static_cast<const uint16_t *>(input_scales);
+
+  // Initialize output to 0
+  for (unsigned int i = 0; i < M * N; ++i) {
+    output[i] = 0.0f;
+  }
+
+  // Input quantization layout parameters
+  unsigned int alignK = (K + quantization_group_size - 1) /
+                        quantization_group_size * quantization_group_size;
+  unsigned int groups_in_row = alignK / quantization_group_size;
+
+  // Outer loop: Height / 32 (N direction)
+  // Blocks of 32 rows (output channels)
+  for (unsigned int n_blk = 0; n_blk < N / 32; ++n_blk) {
+    unsigned int n_start = n_blk * 32;
+
+    // Inner loop: Width / 2 (K direction)
+    // Blocks of 2 columns (input channels)
+    for (unsigned int k_blk = 0; k_blk < K / 2; ++k_blk) {
+      unsigned int k_start = k_blk * 2;
+
+      // Pointer to the 32-byte block of weights
+      // Block index = n_blk * (K / 2) + k_blk
+      // Each block is 32 bytes
+      const uint8_t *w_block = q_weights + (n_blk * (K / 2) + k_blk) * 32;
+
+      // Process for all M rows of input
+      for (unsigned int m = 0; m < M; ++m) {
+        // Load input values and scale
+        // k_start and k_start+1 are in the same group (assuming group_size >=
+        // 2)
+        unsigned int group_id_in_row = k_start / quantization_group_size;
+        unsigned int global_group_id = m * groups_in_row + group_id_in_row;
+
+        // Input scale
+        float i_scale = compute_fp16_to_fp32(in_scales[global_group_id * 2]);
+
+        // Input indices
+        // Note: input is quantized with padding, so we use block addressing
+        unsigned int offset_in_group = k_start % quantization_group_size;
```

Comment:
This does not need to be calculated in this loop. This can be moved out of this loop block to save calculation.